### PR TITLE
BE - implement mark as read

### PR DESCRIPTION
**Pull request by ydayagi**
flpath-666
https://issues.redhat.com/browse/FLPATH-666

### CODE REVIEW

File: plugins/notifications-backend/README.md [hunk 1]
```diff
@@ -59,7 +59,7 @@ User parameter is mandatory because it is needed for message status and filterin
 Request:
 
 ```bash
-curl 'http://localhost:7007/api/notifications/notifications?user=loggedinuser&pageNumber=0&pageSize=0'
+curl 'http://localhost:7007/api/notifications/notifications?user=loggedinuser&read=0&pageNumber=0&pageSize=0'
```
Comment (mareklibra) [inline plugins/notifications-backend/README.md]: Is there a reason to use numbers instead of boolean values (or their strings `"true", "false"` for the `read`? The `readByUser` or the DB is boolean-like.

File: plugins/notifications-backend/src/service/types.ts [hunk 2]
```diff
@@ -36,6 +36,7 @@ export type NotificationsFilter = {
   createdAfter?: Date;
   messageScope?: 'all' | 'user' | 'system';
   user?: string;
+  read?: string; // '0' for unread. '1' for read. undefined for both
```
Comment (mareklibra) [inline plugins/notifications-backend/src/service/types.ts]: We use string for numeric value which is boolean in fact.

File: plugins/notifications-backend/src/service/handlers.ts [hunk 3]
```diff
@@ -239,6 +239,60 @@ export async function getNotificationsCount(
...
+  if (isInsert) {
+    await dbClient('users').insert({
+      message_id: messageID,
+      user: user,
+      read: read,
+    });
+  }
+
+  if (isUpdate) {
+    await dbClient('users')
```
Comment (mareklibra) [inline plugins/notifications-backend/src/service/handlers.ts:289]: As a side-effect (probably not to be fixed now, just to be aware of it), the user can be added among recipients just by knowing the message ID. No matter the user is matching targetGroups  or targetUsers of message creation.

File: plugins/notifications-backend/README.md (hunk 1 above)
Comment (ydayagi) [reply plugins/notifications-backend/README.md]: yes. i could not find a javascript or typescript function that takes a string representation of boolean and returns the boolean value. if you have one then please share. if we can not find one then I think 0/1 is the easiest.

File: plugins/notifications-backend/src/service/types.ts (hunk 2 above)
Comment (ydayagi) [reply plugins/notifications-backend/src/service/types.ts]: the value comes from JSON in HTTP body. so like it or not it is a string. now the question is how do we receive the value from router/express and what does typescript/javascript do for us.
The reality is that they do nothing for us. I get the string from the query object. it can actually be a string array and other types. so i have to put code in router to handle it. now.... typescript does not guarantee that my boolean function var will receive something other than true/false. Si i have to put code in handler as well.
Therefore, my approach is: pass the query object as filter object and do all the handling in handler. in order to compile that the members of filter have to be optional.
Now....since the conversion to boolean from string in javascript is terrible I have to declare the type string and not boolean and do the conversion myself.
Bottom line...it is a string because it is the query parameter and not a converted parameter.

File: plugins/notifications-backend/src/service/handlers.ts (hunk 3 above)
Comment (ydayagi) [reply plugins/notifications-backend/src/service/handlers.ts:289]: of course. security wise i am not concerned. but there could be an accident here. i am waiting for Christophe to reply regarding detecting the requesting/calling user.
I hope we will get rid of the use parameter everywhere.
but even if we do get rid of the user parameter we still have the problem you describe. User A can say: mark messgae-for-user-B as read by A.
We can add validation if you think it is required. i don't mind adding it

Comment (mareklibra) [reply plugins/notifications-backend/src/service/handlers.ts:289]: I would not drive API definition by that.

There is none such converter natively. I believe there will be a few implementations in npm libraries.
Anyway, we can go with something like:
```
    const toBoolean = (str?: string) => {
      if (!str) {
        return undefined;
      }
      return str === 'true' ? true : false;
    }
```

File: plugins/notifications-backend/src/service/router.ts [hunk 4]
```diff
@@ -86,6 +87,36 @@ export async function createRouter(
       .catch(next);
   });
 
+  router.put('/notifications/read', (request, response, next) => {
+    const { messageID, user, read } = request.query;
```
Comment (mareklibra) [inline plugins/notifications-backend/src/service/router.ts]: `messageID` to `messageId`

For consistency, we should change `msgid` in the `createNotification` response to the same `messageId`.

File: plugins/notifications-backend/src/service/router.ts [hunk 5]
```diff
@@ -86,6 +87,36 @@ export async function createRouter(
...
+        break;
+      case '0':
+        readBool = false;
+        break;
+      default:
+        throw new Error('value of parameter "read" must be either "0" or "1"');
+    }
+
+    setRead(dbClient, messageID, user, readBool)
+      .then(result => response.json(result))
```
Comment (mareklibra) [inline plugins/notifications-backend/src/service/router.ts]: We are not sending any data here, the handler's return type is `void`.
We should use
```
.then(() => response.end())
```

File: plugins/notifications-backend/README.md (hunk 1 above)
Comment (mareklibra) [reply plugins/notifications-backend/README.md]: I would not drive API definition by that.

There is none such converter natively. I believe there will be a few implementations in npm libraries.
Anyway, we can go with something like:
```
    const toBoolean = (str?: string) => {
      if (!str) {
        return undefined;
      }
      return str === 'true' ? true : false;
    }
```

File: plugins/notifications-backend/src/service/router.ts (hunk 4 above)
Comment (ydayagi) [reply plugins/notifications-backend/src/service/router.ts]: done

File: plugins/notifications-backend/src/service/router.ts (hunk 5 above)
Comment (ydayagi) [reply plugins/notifications-backend/src/service/router.ts]: done

File: plugins/notifications-backend/README.md (hunk 1 above)
Comment (ydayagi) [reply plugins/notifications-backend/README.md]: done

**Review by mareklibra — APPROVED** [review]
